### PR TITLE
Link bubble menu to preview and edit links

### PR DESCRIPTION
**Pull request by mejo-**
### 📝 Summary

Opens a link bubble when clicking on a link and removes in-document link previews.

* Opens link bubble when clicking on a link or navigating inside a link mark.
* Displays a link preview in the bubble. Loads a fallback if no preview can be loaded.
* Actions inside the bubble: copy link to clipboard, edit link URL, remove link mark.
* Anchor links (href starting with `#`) don't open a link bubble but navigate to the anchor directly.
* No longer display link previews in the document.
* <kbd>Ctrl</kbd> + click and middle click on a link doesn't open link bubble. Instead opens the link directly in `target="_blank"`.
* <kbd>Esc</kbd> closes the link bubble.
* Remove the possibility to customize link click handlers. Instead, we use the reference widgets now, and they should provide their own click handlers if desired.
* Resolves: nextcloud/collectives#845
* Resolves: #2198
* Resolves: #3691

#### 🖼️ Screenshots

| Overview |
|---|
| overview |

Copy link to clipboard | Edit link| Remove mark | Edit mode
--- | --- | --- | ---
clipboard | edit1 | remove | edit

| Screencast |
| --- |
| recording1 |

### 🚧 TODO

* [x] Display link URL if no widget is loaded
* [x] Display link href instead of preview when `!isLoggedIn`
* [x] Read-only support
* [x] Remove link preview from paragraph node view
* [x] Refactor link handling callbacks
* [x] Manual testing
  * [x] Check on mobile
  * [x] Check in public shares
  * [x] Check with references to collective page
* [x] Cypress tests (require #5281 to pass)
* [x] Optional: Display link text in header of link bubble
* [x] Optional: Allow to set text to link title

### 🏁 Checklist

- [x] Code is properly formatted (`npm run lint` / `npm run stylelint` / `composer run cs:check`)
- [x] [Sign-off message](https://probot.github.io/apps/dco/) is added to all commits
- [x] [Tests](https://github.com/nextcloud/text#-testing-the-app) (unit, integration and/or end-to-end) passing and the changes are covered with tests
- [x] Documentation is not required

### CODE REVIEW

**Comment by juliushaertl** [comment]
Looks very nice already. I stumbled over the following two issues that would be worth to check and address in the same run:

- https://github.com/nextcloud/text/issues/2198
- https://github.com/nextcloud/text/issues/3691

File: src/marks/Link.js
Comment (max-nextcloud) [inline src/marks/Link.js:109]: So the `preventDefault()` in `pointerup` is not enough?

That's a pitty as this seems brittle. For example what would happen if i use middle click to open a link in a new tab on windows and then later on try to paste something into a link? Wouldn't `this.storage.nopaste` still be true?

**Review by max-nextcloud — COMMENTED** [review]
The overall approach looks good to me. I added some minor comments but nothing blocking.

File: src/marks/Link.js
Comment (mejo-) [reply src/marks/Link.js:109]: Unfortunately the `preventDefault()` in `pointerup` is not sufficient according to my tests. But you have a fair point. Mhh, have to think more about this.

File: src/components/Link/LinkBubbleView.vue
Comment (mejo-) [reply src/components/Link/LinkBubbleView.vue:73]: Agreed, removed now.

File: src/marks/Link.js
Comment (mejo-) [reply src/marks/Link.js:109]: The best option that comes to my mind is to set `this.storage.nopaste = true` only on Linux (by checking the user agent). What do you think @max-nextcloud?

Comment (mejo-) [reply src/marks/Link.js]: I implemented this now: https://github.com/nextcloud/text/pull/5158/commits/d27d2fa91e421f95cff0abf7f43b5266f4b70b45#diff-90d862ff7cd51c7b18035dbb20c02bc177926bf7224193d0a58b5f870f7a8adbR96-R98

**Comment by mejo-** [comment]
The PR is finally ready for review. Cc @nextcloud/designers.

Cypress tests should pass once #5281 got merged.

**Comment by marcoambrosini** [comment]
Still think that it's a better to allow editing the name within this dialog than relying on editing the string. Mainly because there's ambiguity when it comes to add spaces to the end of the name.

**Comment by juliushaertl** [comment]
I think this is something we should still address differently. We cannot easily just have a textbox for the link title as it may also contain other formatting marks like bold/italic which we could not preserve through a simple input

**Comment by marcoambrosini** [comment]
> may also contain other formatting

I think it's totally fine to discard formatting within the link and use whatever the first character style is for the whole string. Still much less of a drawback than not being able to append a word to the link name IMO. 

Here's an example of google 

https://github.com/nextcloud/text/assets/26852655/64580635-978c-4a55-9eee-fcc006650a1c

**Comment by max-nextcloud** [comment]
> Still much less of a drawback than not being able to append a word to the link name IMO.

I'm not sure why this is not working. I can append a word to an italic text fine - and from the tiptap perspective they are both marks.

**Review by max-nextcloud — APPROVED** [review]

**Comment by juliushaertl** [comment]
The discussion about adding a description field should not block this PR imo, as it would be just another additional new feature. Maybe we can move that to an issue to discuss as a potential follow up.

**Comment by juliushaertl** [comment]
> I'm not sure why this is not working. I can append a word to an italic text fine - and from the tiptap perspective they are both marks.
> update: looks like we inherit this behavior from both tiptap and prosemirror.

Just for reference, quoting prosemirror docs:

> By default, marks are inclusive, meaning that they get applied to content inserted at their end (as well as at their start when they start at the start of their parent node). For link-type marks, this is usually not the expected behavior, and the [inclusive](https://prosemirror.net/docs/ref/#model.MarkSpec.inclusive) property on the mark spec can be set to false to disable that behavior.

And tiptap seems to auto set that if autolink is enabeld
https://github.com/ueberdosis/tiptap/blob/main/packages/extension-link/src/link.ts#L88

**Comment by nimishavijay** [comment]
Agreed with both that the link text can be edited in the popup, and that it can be a follow up PR :) This PR looks really great! Only some minor details for feedback:

- [ ] Indeed the image preview can be smaller, as at this size the text will likely be ellipsized most of the time. How about 64px? :) 
- [ ] There can also be more space between the image preview and the edges of the popup, so some margin can be added
- [ ] The border radius of the image preview should be nicely aligned with border radius of the popup, currently it's too less
- [ ] When you are editing the link, the right and left spacing seems pretty tight. I suggest some left and right margin be added to the text field 
- [ ] In the screenshots there is no text on the left of the buttons, but in the screencast it's there. Is that intentional or am I missing something? I suggest that there doesn't need to be text, but instead there can be an "Open link" secondary button (what do you think @marcoambrosini ?) :)

**Comment by mejo-** [comment]
The failing Cypress test in workspaces.spec.js is because in CI we have `http://localhost:8080` or something alike as local URL and `NcReferenceList` doesn't regard this as a valid URL. The following PR aims to change this: https://github.com/nextcloud-libraries/nextcloud-vue/pull/5176

**Comment by mejo-** [comment]
Thanks for your feedback @marcoambrosini and @nimishavijay.

>  Indeed the image preview can be smaller, as at this size the text will likely be ellipsized most of the time. How about 64px? :)

Given that this would need to happen in the `NcReferenceWidget` component in `nextcloud-vue`, I'd suggest to make this a follow-up topic.

>     In the screenshots there is no text on the left of the buttons, but in the screencast it's there. Is that intentional or am I missing something? I suggest that there doesn't need to be text, but instead there can be an "Open link" secondary button (what do you think @marcoambrosini ?) :)

Addin an "open link" button is a bit more technically involved. The link preview from `NcReferenceWidget` implementations might have custom link handling implemented. Like opening files in the viewer, or opening other pages of a collective via re-routing in vue. Currently there is no easy way to "copy" this behaviour outside the widgets. So I'd rather move this to a follow-up as well.

**Comment by marcoambrosini** [comment]
> what do you think @marcoambrosini

I think that there's no need for such button because there's a link element just below it

**Pull request merged**
